library/core/src/main/java/com/google/android/exoplayer2/text/cea/Cea608Decoder.java

```diff
@@ -307,8 +329,7 @@ protected void decode(SubtitleInputBuffer inputBuffer) {
         continue;
       }
 
-      if ((selectedField == 1 && (ccHeader & CC_FIELD_FLAG) != NTSC_CC_FIELD_1)
-          || (selectedField == 2 && (ccHeader & CC_FIELD_FLAG) != NTSC_CC_FIELD_2)) {
+      if ((ccHeader & CC_FIELD_FLAG) != selectedField) {
```

Comment:
Could you explain a bit more about channels (or provide a link/reference). In particular, why is it that the channel doesn't kick in high up the code (e.g. somewhere around here)? It also seems to not apply to the Basic North American character set on L405, so I'm not sure how channels would work for delivering multiple captions?

Reply:
The standard has lots of ambiguous parts, but it is probably a good start to check paragraph "4.1 Data types in the Line 21 signal" of the 608 standard.
It describes that Field 1 and Field 2 are independent, they are received by observing different parts of the analogue signal.

Each video frame has 2 bytes of data for each Field.

Inside Field 1 there are bytes belonging to CC1, CC2, Text1 and  Text 2.
Inside Field 2 there are bytes belonging to CC3, CC4, Text3, Text4 and XDS.

As the bandwidth is limited, CC1, CC2, Text 1 and Text 2 are sharing the 2 byte per frame data, usually only one service is used completely: CC1 for the primary language. After the commands and characters used for CC1 the leftover of the 60 bytes per second (2 bytes x 30 FPS) can be used for the other services, but it is rare.

Similarly, CC3 of Frame 2 is conveying the second language, CC4, Text3, Text4 and XDS is rarely used.

But when a single frame transports multiple services, we need to check each CC commands which channel it belongs to. And there are additional commands switching to Text an XDS modes. So after those commands all consecutive characters should be put into the buffers of those modes and not into the Caption buffers (this change will come in the next pull request).

Also note that there is only a single bit marking a channel value inside a single frame. So it is sometimes confusing to call the two channels channel 1 and 2, although channel 1 can mean CC1 and CC3 depending on which frame is decoded at this point. Similarly, Channel 2 can mean CC2 and CC4 depending on the current frame.

I cannot quote the standard as it is copyright protected, but I can link some derived wikipedia content:
https://en.wikipedia.org/wiki/EIA-608

> Special North American character set
...
A command to display a special character has a first byte of 0x11 or 0x19 (depending upon channel)

Similarly,
> Extended Western European character set
...
A command to display an extended Spanish/French or miscellaneous character has a first byte of 0x12 or 0x1A (depending upon channel).
...
Portuguese/German/Danish character has a first byte of 0x13 or 0x1B(depending upon channel).

Additionally:
> Control commands
Bits 15 and 7 are always odd parity bits. Bit 11 is always the channel bit. 

The 608 Standard has list of byte values and their meaning, like:
> F.1.1.4 Miscellaneous Control Codes 

```
Data Channel 1      Data Channel 2        Mnemonic Command       Description 
14  20              1C  20                RCL                Resume caption loading 
```
So the good news is that we always need to check a single bit of the commands and based on that bit we can decide which channel the command belongs to. (The only exceptions are MISC codes).

All consecutive characters not having a channel bit should be directed into the last used channel. So if there is a command to switch to "White Underlined Font" on Channel 1, we need to check the frame value to decide if that is CC1 or CC3. And from that point on all incoming characters will belong to that caption channel.

MISC codes also have one additional bit expected to change based on the Field value. It is described in this very convoluted way in the standard:
>8.4 Closed Caption Mode 
The non-printing character of the miscellaneous control-character pairs that fall in the range of 0x14, 0x20 to 0x14, 0x2F in field 1, shall be replaced with 0x15, 0x20 to 0x15, 0x2F when used in field 2.  
b)   The non-printing character of the miscellaneous control-character pairs that fall in the range of 0x1C, 0x20 to 0x1C, 0x2F in field 1, shall be replaced with 0x1D, 0x20 to 0x1D, 0x2F when used in field 2.  
c)   When any of the control codes from 0x01 to 0x0F is used to begin a control-character pair it indicates the beginning of XDS Data. This XDS data should be 
considered part of a different data channel and can be used to interrupt the captioning data according to the FCC rules for other data channels.